### ISSUE TITLE
Google Consent Mode changes form action on submit to include [object%20HTMLInputElement]?gl=123

### ISSUE THREAD

**Issue by joepagan** [verbb/formie#2159]
### Describe the bug

On one of our apps which uses formie, when a user has only opted for necessary cookies and later submits a form they will be redirected to a url like this (even when ajax submission is on the formie form).

I notice the action changes on the submit event.

This looks to be intention as a way for Google to send a cookieless ping to google however as a concequence it redirects the user to the action:

```sh
domain.com/[object%20HTMLInputElement]?_gl=1*1trac45*_up*MQ..*_ga*NDE0NTIzMDMwLjE3MzI1NzAzMDg.*_ga_VGE67CL4VC*MTczMjU3MDMwOC4xLjEuMTczMjU3MDMxNC4wLjAuMA..
```

It looks like Brandon has come up with a suggestion:
https://craftcms.stackexchange.com/questions/41657/url-passthrough-rewrite-failing-when-using-actioninput

Looks like other devs have reported the same issue to Google who will of course never respond:
https://support.google.com/tagmanager/thread/269636163/form-action-fields-changes-when-form-is-submitted-after-applying-consent-mode-v2?hl=en

However, I wonder if it's possible for the formie JS to somehow detect an action on a form element and prevent a redirect onwards, especially if the form is in ajax submission mode?

FYI - it does look like the submission data is stored, however they appear to have been labelled as spam in the backend.

### Steps to reproduce

1. have consent mode v2
2. opt out of non-essential cookies
3. submit formie form
4. get redirect to page url like `/[object%20HTMLInputElement]?_gl=x`

### Form settings

- Multi-page form: Yes or No
- Submission Method: Ajax or Page Reload
- Client-side Validation: Yes or No
- Custom Form Templates: Yes or No

### Craft CMS version

5

### Plugin version

3

### Multi-site?

yes

**Comment by engram-design** [comment]
Very interesting, first I've heard about this!

If it helps, the same workaround can work in Formie with Theme Config:

```twig
{{ craft.formie.renderForm(form, {
    themeConfig: {
        form: {
            attributes: {
                action: craft.app.request.absoluteUrl,
            },
        },
    },
}) }}
```

**Comment by joepagan** [comment]
Hey @engram-design yes I did try exactly this suggestion, however despite being on an ajax formie form, submitting with a custom action will cause a redirect to the action URL rather than preventing it and showing the success message.

**Comment by engram-design** [comment]
@joepagan Just tested that with an Ajax form, and that's working as expected to me, sorry to say. It's not redirecting to the actionUrl. There's no JS errors when submitting the form?

Here's how we [handle it](https://github.com/verbb/formie/blob/d0da4a6e83c13ae9e1f5dee2c36d0e769495fc20/src/web/assets/frontend/src/js/formie-form-theme.js#L415).

**Comment by Tim-Wils** [comment]
I stumbled on this a few months back and I discovered that Google is adding the form action-attribute where Formie is not adding an action-attribute, but rather uses the input[hidden][name='action'] input for the action.

So I figured I just have to remove the action-attribute on the form-element whenever the name='action' input exists and it's added by Google.

Here's my code whichs reads at: if there's a <input type='hidden'  name='action'> present, observe the form for the action attribute and remove it, whenever it's added.

Hope it helps someone:
```
<script>
const formobserver = new MutationObserver(mutations => {
  mutations.forEach(mutation => {
    if (mutation.type === 'attributes' && mutation.attributeName === 'action') {
      mutation.target.removeAttribute('action');
    }
  });
});

const forms = document.querySelectorAll('form[data-fui-form]');
for (let i = 0; i < forms.length; i++) {
	const element = forms[i];
	if (element.querySelector('input[name="action"]')) {
		formobserver.observe(element,{ attributes: true })
	}
}
</script>
```

**Comment by joepagan** [comment]
Just to update this, it appears this appears to occur when users only opt-in to essential cookies. Thanks for the suggestion @Tim-Wils I'll update when we've given that a go

**Comment by Tim-Wils** [comment]
@joepagan I know :) Honestly, this was a quick-fix to solve it ASAP for our client. We still don't know how to solve this in a neat way though. 
Suggestions/settings in GA4/GAM to solve this are welcome.

**Comment by joepagan** [comment]
Here is my progress on this:
Unfortunately the suggested change does not resolve the issue for me. 
- The script detects the mutation, and removes the attribute after submission, but it looks like the POST is called before it's able to remove it.
- A native `submit` event handler and the formie `onBeforeFormieSubmit` event is called before this mutation occurs and the POST request is triggered.
- Using a JS watch event for requests being made (similar to the mutation above) does not appear to fire for this POST request.
- I have found that it looks like I can no longer replicate the issue on Ajax form submissions, however page-reload formie forms are still behaving badly.
- Turning off advanced tracking on form submissions in GA4 (in 2 places) does not change this.
- Replacing the formie form with a native does not improve this
- I have also found [another locked (😠 why lock it when there is no answer ffs) google community post](https://support.google.com/google-ads/thread/173640313/url-passthrough-gclid-is-breaking-form-actions?hl=en) about this issue so I wonder if we could make a change to formie so it honours an action attribute on the form element instead of the hidden input (and remove the hidden input) this may solve it.
- Updating the form action attribute to use `action="{{ actionUrl('users/login') }}"` and removing the hidden input does not work, it also creates a browser prompt as actionUrl includes the domain which I guess the browser interprets as a separate domain and hints you are leaving the current page.
- At one point I thought it might be related to my password manager, which once upon a time stored values of hidden inputs in forms, setting a blank value for `action` in the password manager appeared to work for a short while, however, testing in chrome > guest mode where no extensions exist where I enter the values manually I can still reproduce this 😭 .
- Updating the form action to use `/actions/users/login` and removing the hidden input WORKS logging in with the main admin account, however a frontend only user account on the POST to `/actions/users/login?_gl=...` will 404, what the actual hell...

**Comment by joepagan** [comment]
As an update, I _may_ have resolved this. For now I cannot replicate it, I'll keep this open as I'm not confident behaviour may change right now and maybe GA changes take a little while to come through. Though I'll try update within the next day & close this

- I updated my GTM tags, paused the legacy UA tag, updated consent settings against each tag which were missing
- paused ga4 - problem went away
- resumed ga4 - problem came back

To confirm, the form element does not have an `action` attribute and just uses the hidden input like is expected.

The thing that I _think_ that made a difference:
by updating the settings in GA4 > admin > data streams > website > enhanced measurement.
We had form interactions disabled (which I would have thought would be the only measurement which may cause this), though what is weird is that I don't think I actually changed anything... I turned off file downloads as a test, but I don't believe I hit the save button and clicked off the modal (closing the enhanced measurement UI). Though maybe turning a setting on/off may have reset saving these values and maybe google had some different saved values in their DB.

**Comment by engram-design** [comment]
Really appreciate all the updates on this @joepagan it's all stuff I've not had the _pleasure_ of dealing with yet in projects.

**Comment by joepagan** [comment]
I _thought_ we may have cracked this but unfortunately my colleague could replicate it last week. 
Strangely, they could only replicate the issue with a specific Craft User!
Deleting the Craft user and re-creating them seemed to resolve it, not sure why as we don't store redirect URLs against users or anything like that... 
It sounds like some odd caching however all caches have been cleared after multiple deployments testing this so I don't believe it was caching related.
I'll see if time proves that this has resolved it and if so I'll highlight to the pixel and tonic peeps.

**Comment by Tim-Wils** [comment]
That's odd, because we always work with a fully statically cached pages and the action is added bij GTM with js. (so after I observe the addition of the false action, I remove it)
For me to replicate this behaviour consistently is:
- decline the cookieconsent
- then the GTM/GA4 trick to persistently track everything kicks in: they put their weird parameters after every url / form action / button wherever the want to

**Comment by joepagan** [comment]
>so after I observe the addition of the false action, I remove it

I don't think this every worked for us because the action was changed by something in the google script _after_ submission, any event which watched for manipulation occurred after the submission and it was too late by then to do anything about it! 

I haven't seen this for a really long time now luckily, hoping it never comes back, I'm going to close this now.
